bundles/framework/featuredata/plugin/FeatureDataPluginHandler.js

```diff
@@ -170,18 +193,148 @@ class FeatureDataPluginUIHandler extends StateHandler {
+        }
+        return settings?.allColumns ? { allColumns: settings.allColumns, filters, errors: null } : null;
+    }
+
     closeFlyout (resetLayers) {
         if (this.flyoutController) {
             this.flyoutController.close();
             this.flyoutController = null;
         }
-        this.updateState({ flyoutOpen: false, layers: resetLayers ? null : this.getState().layers });
+
+        if (this.popupController) {
+            this.popupController.close();
+            this.popupController = null;
+        }
+
+        this.updateState({
+            flyoutOpen: false,
+            layers: resetLayers ? null : this.getState().layers
+        });
     }
 
     updateFlyout () {
         if (this.flyoutController) {
             this.flyoutController.update(this.getState());
         }
+
+        if (this.popupController) {
+            this.popupController.update(this.getState());
+        }
+    }
+
+    updateFilters (index, filter) {
+        const { selectByPropertiesSettings } = this.getState();
+        selectByPropertiesSettings.filters[index] = filter;
+        filter.error = null;
+        this.updateState({ selectByPropertiesSettings });
+    }
+
+    applyFilters () {
+        const { selectByPropertiesSettings } = this.getState();
+        const { filters } = selectByPropertiesSettings;
+
+        let hasErrors = false;
+        filters.forEach((filter) => {
+            if (!filter?.value?.length) {
+                hasErrors = true;
+                filter.error = true;
+            }
+        });
+
+        if (hasErrors) {
+            this.updateState({ selectByPropertiesSettings });
+            return;
+        }
+
+        this.selectFeaturesByProperties();
+    }
+
+    selectFeaturesByProperties () {
+        const { selectByPropertiesSettings, activeLayerId, activeLayerFeatures } = this.getState();
+        const filters = selectByPropertiesSettings.filters;
+
+        // separate the logical operators from the original filter and form a single array
+        const filterArray = [];
+        filters.forEach((filter) => {
+            filterArray.push(filter);
+            if (filters.length > 1) {
+                filterArray.push({ boolean: filter.logicalOperator });
+            }
+        });
+
+        const filteredIds = new Set();
+        const alternatives = getFilterAlternativesAsArray(filterArray);
+        alternatives.forEach(attributeFilters => {
+            let filteredList = activeLayerFeatures;
+            attributeFilters.forEach(filter => {
+                filteredList = filterFeaturesByAttribute(filteredList, filter);
+            });
+            filteredList
+                .map((feature) => feature.id)
+                .filter(id => !!id)
+                .forEach(id => filteredIds.add(id));
+        });
+
+        this.selectionService.setSelectedFeatureIds(activeLayerId, [...filteredIds]);
+        this.updateState({ selection: [...filteredIds] });
```

Comment:
We should remove updateState() from here and do it when we receive the event about selected features. That way we don't need to care how the features got selected and selecting from map works as well as selecting through this tool.

Reply:
Removed excessive state update.